tap_rest_api_msdk/streams.py

```diff
@@ -1,13 +1,184 @@
 """Stream type classes for tap-rest-api-msdk."""
 
 from datetime import datetime
-from typing import Any, Dict, Iterable, Optional
+from typing import Any, Dict, Iterable, Optional, cast
 
 import requests
+import email.utils
+from singer_sdk.pagination import SinglePagePaginator, BaseOffsetPaginator, BaseHATEOASPaginator, JSONPathPaginator, HeaderLinkPaginator, SimpleHeaderPaginator, BasePageNumberPaginator
 from singer_sdk.helpers.jsonpath import extract_jsonpath
 from tap_rest_api_msdk.client import RestApiStream
 from tap_rest_api_msdk.utils import flatten_json
-from urllib.parse import urlparse, parse_qsl
+from urllib.parse import urlparse, parse_qsl, parse_qs
+
+
+class RestAPIBasePageNumberPaginator(BasePageNumberPaginator):
+    def __init__(
+        self,
+        *args,
+        jsonpath: str = None,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self._jsonpath = jsonpath
+
+    def has_more(self, response: requests.Response):
+        """Return True if there are more pages to fetch.
+
+        Args:
+            response: The most recent response object.
+            jsonpath: An optional jsonpath to where the tokens are located in
+                      the response, defaults to `hasMore` in the response.
+
+        Returns:
+            Whether there are more pages to fetch.
+        """
+        
+        if self._jsonpath:
+            return next(extract_jsonpath(self._jsonpath, response.json()), None)
+        else:
+            return response.json().get("hasMore", None)
+
+class RestAPIOffsetPaginator(BaseOffsetPaginator):
+    def __init__(
+        self,
+        *args,
+        jsonpath: str = None,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self._jsonpath = jsonpath
+
+    def has_more(self, response: requests.Response):
+        """Return True if there are more pages to fetch.
+
+        Args:
+            response: The most recent response object.
+            jsonpath: An optional jsonpath to where the tokens are located in
+                      the response, defaults to pagination in the response.
+
+        Returns:
+            Whether there are more pages to fetch.
+        """
+        
+        if self._jsonpath:
+            pagination = next(extract_jsonpath(self._jsonpath, response.json()), None)
+        else:
+            pagination = response.json().get("pagination", None)
+
+        if pagination and all(x in pagination for x in ["offset", "limit"]):
+            record_limit = pagination.get("total",pagination.get("count",0))
```

Comment:
I like the use of zero here if there isn't a count or total value, meaning the tap will stop

Reply:
dbt API wouldn't work with this, but I guess you can't cater for everything. In dbt the limit and offset parameters fall under the `filters` key, not the `pagination` key.

```json
"extra":{"filters":{"limit":15,"offset":204,"account_id":1},"order_by":"id","pagination":{"count":15,"total_count":219}}}
```

Reply:
Oh, you should be able to work with the dbt api. Just add the jsonpath to that location e.g.  `TAP_REST_API_MSDK_NEXT_PAGE_TOKEN_PATH='$.extra.filters'` .

Good suggestion re stopping if no total or count is present.

Reply:
Maybe I need to document this a bit better in my comments for this function.

Reply:
Just adding, as discussed, it may be possible to get pagination values out of the dbt Cloud API using some sort of JSONPath expression like this:

`$.extra.[filters,pagination]`

This is required because the `limit` and `offset` keys appear under `filters` and the `count` and `total_count` keys appear in a different place, under the `pagination` key:

```json
{
"extra": {
    "filters": {
        "limit": 100,
        "offset": 0,
        "account_id": 1
    },
    "order_by": "id",
    "pagination": {
        "count": 10,
        "total_count": 10
    }
}
}
```

Reply:
Thanks for providing that example. I have added an additional function `compress_dict` which will post-process the pagination dictionary to create a single dictionary of keypairs thus removing the nesting causing this issue. The token can now be updated to `TAP_REST_API_MSDK_NEXT_PAGE_TOKEN_PATH='$.extra'` . I will add a test set of config for dbt cloud to the README.md file.

Reply:
I should note that the key `total_count` is not the default key of `total`. To change the total limit key checked to be `total_count` an additional parameter must be set `TAP_REST_API_MSDK_PAGINATION_TOTAL_LIMIT_PARAM="total_count"`.